auth/idtoken/file.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package idtoken
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/auth"
+	"cloud.google.com/go/auth/detect"
+	"cloud.google.com/go/auth/impersonate"
+	"cloud.google.com/go/auth/internal/internaldetect"
+)
+
+const (
+	jwtTokenURL = "https://oauth2.googleapis.com/token"
+	iamCredAud  = "https://iamcredentials.googleapis.com/"
+)
+
+var (
+	defaultScopes = []string{
+		"https://iamcredentials.googleapis.com/",
+		"https://www.googleapis.com/auth/cloud-platform",
+	}
+)
+
+func tokenProviderFromBytes(b []byte, opts *Options) (auth.TokenProvider, error) {
+	t, err := internaldetect.ParseFileType(b)
+	if err != nil {
+		return nil, err
+	}
+	switch t {
+	case internaldetect.ServiceAccountKey:
+		f, err := internaldetect.ParseServiceAccount(b)
+		if err != nil {
+			return nil, err
+		}
+		opts2LO := &auth.Options2LO{
```

Comment:
I really like the explicit options type for "2LO" here, this is much easier to follow than the old design.

Reply:
:)